### PR TITLE
kata: introduce kata container support

### PR DESCRIPTION
**Pull request by egernst**
Signed-off-by: Eric Ernst <eric.ernst@intel.com>

**What this PR does / why we need it**:

PR adds support for Kata Containers. 

**Which issue this PR fixes** : 

fixes # 3463

**Special notes for your reviewer**:

Should look very familiar to initial Clear Container support PRs.

**If applicable**:
- [x] documentation
- [x] unit tests
 
**Release note**:

```release-note
```

### CODE REVIEW

**Comment by egernst** [comment]
testing still wip

**Comment by egernst** [comment]
In my testing, I see:

```
eernst-mac02:acs-engine eernst$ ./bin/acs-engine deploy --subscription-id 052907a4-eab1-46e9-937c-af13b410c521 --dns-prefix kata-test --location westus2 --api-model kubernetes-kata-containers.json
WARN[0003] apimodel: missing masterProfile.dnsPrefix will use "kata-test" 
WARN[0003] --resource-group was not specified. Using the DNS prefix from the apimodel as the resource group name: kata-test 
WARN[0006] apimodel: ServicePrincipalProfile was missing or empty, creating application... 
WARN[0008] created application with applicationID (ef542ac5-28c0-45c0-bbf4-f8eefec6a701) and servicePrincipalObjectID (2d64c93b-de51-4801-ba8b-9ab2bc3e5c55). 
WARN[0008] apimodel: ServicePrincipalProfile was empty, assigning role to application... 
INFO[0045] Starting ARM Deployment (kata-test-1196995142). This will take some time... 
INFO[0259] Finished ARM Deployment (kata-test-1196995142). Error: resources.DeploymentsClient#CreateOrUpdate: Failure sending request: StatusCode=200 -- Original Error: Long running operation terminated with status 'Failed': Code="DeploymentFailed" Message="At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-debug for usage details." 
ERRO[0259] {"status":"Failed","error":{"code":"DeploymentFailed","message":"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-debug for usage details.","details":[{"code":"Conflict","message":"{\r\n  \"status\": \"Failed\",\r\n  \"error\": {\r\n    \"code\": \"ResourceDeploymentFailure\",\r\n    \"message\": \"The resource operation completed with terminal provisioning state 'Failed'.\",\r\n    \"details\": [\r\n      {\r\n        \"code\": \"VMExtensionProvisioningError\",\r\n        \"message\": \"VM has reported a failure when processing extension 'cse-agent-0'. Error message: \\\"Enable failed: failed to execute command: command terminated with exit status=2\\n[stdout]\\n\\n[stderr]\\n\\\".\"\r\n      }\r\n    ]\r\n  }\r\n}"},{"code":"Conflict","message":"{\r\n  \"status\": \"Failed\",\r\n  \"error\": {\r\n    \"code\": \"ResourceDeploymentFailure\",\r\n    \"message\": \"The resource operation completed with terminal provisioning state 'Failed'.\",\r\n    \"details\": [\r\n      {\r\n        \"code\": \"VMExtensionProvisioningError\",\r\n        \"message\": \"VM has reported a failure when processing extension 'cse-agent-2'. Error message: \\\"Enable failed: failed to execute command: command terminated with exit status=2\\n[stdout]\\n\\n[stderr]\\n\\\".\"\r\n      }\r\n    ]\r\n  }\r\n}"},{"code":"Conflict","message":"{\r\n  \"status\": \"Failed\",\r\n  \"error\": {\r\n    \"code\": \"ResourceDeploymentFailure\",\r\n    \"message\": \"The resource operation completed with terminal provisioning state 'Failed'.\",\r\n    \"details\": [\r\n      {\r\n        \"code\": \"VMExtensionProvisioningError\",\r\n        \"message\": \"VM has reported a failure when processing extension 'cse-agent-1'. Error message: \\\"Enable failed: failed to execute command: command terminated with exit status=2\\n[stdout]\\n\\n[stderr]\\n\\\".\"\r\n      }\r\n    ]\r\n  }\r\n}"},{"code":"Conflict","message":"{\r\n  \"status\": \"Failed\",\r\n  \"error\": {\r\n    \"code\": \"ResourceDeploymentFailure\",\r\n    \"message\": \"The resource operation completed with terminal provisioning state 'Failed'.\",\r\n    \"details\": [\r\n      {\r\n        \"code\": \"VMExtensionProvisioningError\",\r\n        \"message\": \"VM has reported a failure when processing extension 'cse-master-0'. Error message: \\\"Enable failed: failed to execute command: command terminated with exit status=2\\n[stdout]\\n\\n[stderr]\\n\\\".\"\r\n      }\r\n    ]\r\n  }\r\n}"}]}} 
FATA[0259] resources.DeploymentsClient#CreateOrUpdate: Failure sending request: StatusCode=200 -- Original Error: Long running operation terminated with status 'Failed': Code="DeploymentFailed" Message="At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-debug for usage details." 
```

**Comment by egernst** [comment]
@scooley, @jessfraz  - let me know if this failure is familiar, or if you have tips on where to look for more details. Thx!

File: parts/k8s/kubernetescustomscript.sh [hunk 1]
```diff
@@ -344,13 +364,13 @@ function installContainerd() {
 	sed -i '/\[Service\]/a ExecStartPost=\/sbin\/iptables -P FORWARD ACCEPT' /etc/systemd/system/containerd.service
 
 	echo "Successfully installed cri-containerd..."
-	if [[ "$CONTAINER_RUNTIME" == "clear-containers" ]] || [[ "$CONTAINER_RUNTIME" == "containerd" ]]; then
+	if [[ "$CONTAINER_RUNTIME" == "clear-containers" ]] [[ "$CONTAINER_RUNTIME" == "kata-containers" ]] || [[ "$CONTAINER_RUNTIME" == "containerd" ]]; then
```
Comment (adelina-t) [inline parts/k8s/kubernetescustomscript.sh]: Forgot "||" operator here :)

File: parts/k8s/kubernetescustomscript.sh [hunk 2]
```diff
@@ -344,13 +364,13 @@ function installContainerd() {
...
 		setupContainerd
 	fi
 }
 
 function ensureContainerd() {
-	if [[ "$CONTAINER_RUNTIME" == "clear-containers" ]] || [[ "$CONTAINER_RUNTIME" == "containerd" ]]; then
+	if [[ "$CONTAINER_RUNTIME" == "clear-containers" ]] [[ "$CONTAINER_RUNTIME" == "kata-containers" ]] || [[ "$CONTAINER_RUNTIME" == "containerd" ]]; then
```
Comment (adelina-t) [inline parts/k8s/kubernetescustomscript.sh]: and here :) that's why you got cse-master-0 error.

**Comment by egernst** [comment]
@adelina-t - sweet, thanks for teaching me  to fish - the /var/log/azure/cluster-provision.log  is very helpful; that's what I was missing.

File: parts/k8s/kubernetescustomscript.sh (hunk 2 above)
Comment (egernst) [reply parts/k8s/kubernetescustomscript.sh]: Oh, that's embarrassing -- thanks for the help.

**Comment by egernst** [comment]
@adelina-t  -- repushed with these fixed.  

I verified I could create a cluster using the added example, ```kubernetes-kata-containers.json```

You mentioned a comment inline?  Can you clarify.

File: examples/e2e-tests/kubernetes/kubernetes-config/kata-containers.json [hunk 3]
```diff
@@ -0,0 +1,40 @@
+{
+  "apiVersion": "vlabs",
```
Comment (CecileRobertMichon) [inline examples/e2e-tests/kubernetes/kubernetes-config/kata-containers.json]: I don't think we need the example json in both places, we have a //todo item to move the feature examples in `e2e-tests/` so you can probably keep just the other one

**Comment by adelina-t** [comment]
@egernst  By "comment inline" I was referring to the comments I left on the commit. Probably not the clearest choice of words on my part :)

File: parts/k8s/kubernetescustomscript.sh [hunk 4]
```diff
@@ -281,6 +281,24 @@ function configNetworkPlugin() {
     fi
 }
 
+function installKataContainersRuntime() {
+    # Add Kata Containers repository key
+    echo "Adding Kata Containers repository key..."
+    KATA_RELEASE_KEY_TMP=/tmp/kata-containers-release.key
+    KATA_URL=http://download.opensuse.org/repositories/home:/katacontainers:/release/xUbuntu_16.04/Release.key
+    retrycmd_if_failure_no_stats 20 1 5 curl -fsSL $KATA_URL > $KATA_RELEASE_KEY_TMP || exit $ERR_APT_INSTALL_TIMEOUT
```
Comment (CecileRobertMichon) [inline parts/k8s/kubernetescustomscript.sh]: let's add a new exit code specific to Kata install here

File: parts/k8s/kubernetescustomscript.sh [hunk 5]
```diff
@@ -281,6 +281,24 @@ function configNetworkPlugin() {
...
+    retrycmd_if_failure 10 5 10 apt-key add $KATA_RELEASE_KEY_TMP || exit $ERR_APT_INSTALL_TIMEOUT
+
+    # Add Kata Container repository
+    echo "Adding Kata Containers repository..."
+    echo 'deb http://download.opensuse.org/repositories/home:/katacontainers:/release/xUbuntu_16.04/ /' > /etc/apt/sources.list.d/kata-containers.list
+
+    # Install Kata Containers runtime
+    echo "Installing Kata Containers runtime..."
+    apt_get_update
```
Comment (CecileRobertMichon) [inline parts/k8s/kubernetescustomscript.sh]: this should be `apt_get_update || exit $ERR_APT_UPDATE_TIMEOUT`

File: parts/k8s/kubernetescustomscript.sh [hunk 6]
```diff
@@ -281,6 +281,24 @@ function configNetworkPlugin() {
...
+    apt_get_install 20 30 120 kata-runtime
```
Comment (CecileRobertMichon) [inline parts/k8s/kubernetescustomscript.sh]: and `apt_get_install 20 30 120 kata-runtime  || exit $ERR_KATA_INSTALL_TIMEOUT` here, ERR_KATA_INSTALL_TIMEOUT needs to be defined at the top of the file

File: parts/k8s/kubernetescustomscript.sh (hunk 4 above)
Comment (CecileRobertMichon) [reply parts/k8s/kubernetescustomscript.sh]: You can see `installDocker()` for a model

File: parts/k8s/kubernetescustomscript.sh [hunk 7]
```diff
@@ -281,6 +281,24 @@ function configNetworkPlugin() {
     fi
 }
 
+function installKataContainersRuntime() {
```
Comment (CecileRobertMichon) [inline parts/k8s/kubernetescustomscript.sh]: I realize that my comments in this function also apply to `installClearContainersRuntime()`, which is probably why you did it that way but I think `installClearContainersRuntime` needs to be changed as well to have proper exit codes.

File: parts/k8s/kubernetescustomscript.sh [hunk 8]
```diff
@@ -546,6 +566,14 @@ if [[ "$CONTAINER_RUNTIME" == "clear-containers" ]]; then
 		installClearContainersRuntime
 	fi
 fi
+
+if [[ "$CONTAINER_RUNTIME" == "kata-containers" ]]; then
+	# Ensure we can nest virtualization
+	if grep -q vmx /proc/cpuinfo; then
```
Comment (CecileRobertMichon) [inline parts/k8s/kubernetescustomscript.sh]: what happens if this is false?

File: pkg/api/vlabs/validate_test.go [hunk 9]
```diff
@@ -1083,6 +1083,18 @@ func Test_Properties_ValidateContainerRuntime(t *testing.T) {
 		)
 	}
 
+	p.OrchestratorProfile.KubernetesConfig.ContainerRuntime = "kata-containers"
```
Comment (CecileRobertMichon) [inline pkg/api/vlabs/validate_test.go]: Is kata-containers supported for all k8s versions?

File: examples/e2e-tests/kubernetes/kubernetes-config/kata-containers.json (hunk 3 above)
Comment (egernst) [reply examples/e2e-tests/kubernetes/kubernetes-config/kata-containers.json]: ACK.

File: pkg/api/vlabs/validate_test.go (hunk 9 above)
Comment (egernst) [reply pkg/api/vlabs/validate_test.go]: Kata is more tightly couple with the CRI-shim version (in this case, containerd).  I think if there's an error, it'll likely be a mismatch between containerd + k8s?

File: parts/k8s/kubernetescustomscript.sh (hunk 7 above)
Comment (egernst) [reply parts/k8s/kubernetescustomscript.sh]: No worries.  Okay to update the example function, `installClearContainersRuntime()`, in a follow-on PR?

File: parts/k8s/kubernetescustomscript.sh (hunk 5 above)
Comment (egernst) [reply parts/k8s/kubernetescustomscript.sh]: ack

File: parts/k8s/kubernetescustomscript.sh (hunk 6 above)
Comment (egernst) [reply parts/k8s/kubernetescustomscript.sh]: ack

File: parts/k8s/kubernetescustomscript.sh (hunk 4 above)
Comment (egernst) [reply parts/k8s/kubernetescustomscript.sh]: ACK.

File: parts/k8s/kubernetescustomscript.sh (hunk 8 above)
Comment (egernst) [reply parts/k8s/kubernetescustomscript.sh]: in that case, the given node would not install kata container artifacts

File: examples/kubernetes-kata-containers.json [hunk 10]
```diff
@@ -0,0 +1,50 @@
+{
+  "apiVersion": "vlabs",
+  "properties": {
+    "orchestratorProfile": {
+      "orchestratorType": "Kubernetes",
+      "orchestratorRelease": "1.10",
+      "kubernetesConfig": {
+        "networkPlugin": "flannel",
+        "containerRuntime": "kata-containers",
+        "addons": [
```
Comment (CecileRobertMichon) [inline examples/kubernetes-kata-containers.json]: any reason why tiller and dashboard are disabled explicitly? If not, we can probably let them go to default

File: parts/k8s/kubernetescustomscript.sh (hunk 8 above)
Comment (egernst) [reply parts/k8s/kubernetescustomscript.sh]: It'll be fully functional, though the user may hope to use Kata, but in reality being using `runc`. 
When Kata is installed, the operator deploying workloads would have option of either using runc or kata-runtime.  In the case VMX isn't supported on the node, any workloads targeting `kata-runtime` would be handled by the default runtime, `runc`.

File: examples/kubernetes-kata-containers.json (hunk 10 above)
Comment (egernst) [reply examples/kubernetes-kata-containers.json]: No, this was just me being lazy and copying what was done for Clear Containers in the past.

File: parts/k8s/kubernetescustomscript.sh (hunk 8 above)
Comment (CecileRobertMichon) [reply parts/k8s/kubernetescustomscript.sh]: Sounds good, thanks for clarifying

**Comment by egernst** [comment]
@CecileRobertMichon I  verified example after suggested lines. PTAL

**Comment by egernst** [comment]
I'm not sure if there is any action required on my end for kicking this CI (waiting for status to be reported), but please let me know if I can do anything to help push this along.

**Pull request merged**